### PR TITLE
Bug fix proposal: Handle errors appearing when trying to close closed…

### PR DESCRIPTION
**Pull request by ksallberg**
… (or nonexisting ports). Also avoid sending stop call to verx_client_unix gen_server when receiving EXIT info from libvirtd, as this causes another error. Fixed typo in README.

I'm doing tests where I launch and remove around 10 Ubuntu VMs using verx. In these tests, I have encountered two errors:

(1) As libvirtd sends EXIT (line 147 of verx_client_unix.erl), previously an attempt to shut down the affected gen_server was made. This results the following errors: {{shutdown,normal} ,{gen_server,call,[<0.180.0>,stop]}}. Do you agree that when libvirtd sends EXIT, the process stops anyway, and therefore stopping it again yields this error? Or am I mistaken? This error is rare and happens ~1 time per all calls made when starting 10 VMs. I ran length(erlang:processes()) at times between creating and destroying sets of 10 VMs, it seems the total amount of processes stays stable when not stopping the gen_servers as before.

(2) Fairly often, when calling erlang:port_close (line 155/156 of verx_client_unix.erl), a badarg error is raised. I have no idea on how to solve this, but I suggest that we catch this error in the mean time. I could look further into this if you have any ideas on why this might be.

### CODE REVIEW

**Comment by msantos** [comment]
On Thu, Jul 30, 2015 at 07:44:00AM -0700, Kristian Sällberg wrote:
> … (or nonexisting ports). Also avoid sending stop call to verx_client_unix gen_server when receiving EXIT info from libvirtd, as this causes another error. Fixed typo in README.
> 
> I'm doing tests where I launch and remove around 10 Ubuntu VMs using verx. In these tests, I have encountered two errors:
> 
> (1) As libvirtd sends EXIT (line 147 of verx_client_unix.erl), previously an attempt to shut down the affected gen_server was made. This results the following errors: {{shutdown,normal} ,{gen_server,call,[<0.180.0>,stop]}}. Do you agree that when libvirtd sends EXIT, the process stops anyway, and therefore stopping it again yields this error? Or am I mistaken? This error is rare and happens ~1 time per all calls made when starting 10 VMs. I ran length(erlang:processes()) at times between creating and destroying sets of 10 VMs, it seems the total amount of processes stays stable when not stopping the gen_servers as before.

The 'EXIT' you are seeing is being sent by the port holding the fd of
the TCP connection to libvirtd. What I suspect is happening:

* libvirtd closes the socket

* the Erlang port sees the socket has closed and sends an exit with reason
  'normal'

* since handle_info/2 returns a tuple in stop ({stop, {shutdown,normal}, State},
  the gen_server prints out an error message

So there are a few problems here:

* the man page for erlang:open_port/2 says:

        During  use  of a port opened using {spawn, Name}, {spawn_driver,
        Name} or {spawn_executable, Name}, errors arising when sending
        messages to it are reported to the owning process using signals
        of the form {'EXIT', Port, PosixCode}. See file(3) for possible
        values of PosixCode.

  So the intent here was to stop the gen_server and log errno but that
  does not work because we are using {fd, FD, FD} to open the port. Or,
  possibly since errno was not set in this case, the port sends the atom
  'normal'.

  If we shutdown the gen_server using 'shutdown' or 'normal', it should
  suppress the error messages:

        {stop, shutdown, State}

* my guess for why you are not seeing the number of processes grow is
  that the next call that uses the port will terminate the gen_server with
  badarg

> (2) Fairly often, when calling erlang:port_close (line 155/156 of verx_client_unix.erl), a badarg error is raised. I have no idea on how to solve this, but I suggest that we catch this error in the mean time. I could look further into this if you have any ideas on why this might be.

Looks good! One small problem: closing the fd in the try block means it
will not be called if the `erlang:port_close(Port)` fails, so the
gen_server will end up leaking fd's. We just need to move the
`procket:close(Fd)` after the try/catch.

The badarg occurs because the gen_server is trying to close an invalid
port. The port was closed by the runtime when we got the 'EXIT' message.

It would be interesting to figure out what is going on. Do you see
anything in the libvirtd logs? If not, can you run libvirtd in debug
mode and repeat your tests?

I also see another fd leak around line 68 in the init/1 function:

~~~ erlang
ok = procket:connect(Socket, Sun),
~~~

If connect/2 fails, the gen_server process will crash. I don't think
terminate/2 will be called in this situation so the fd will leak. Maybe the
appropriate thing to do here is error out:

~~~ erlang
case procket:connect(Socket, Sun)
    ok ->
        Port = erlang:open_port({fd, Socket, Socket}, [
                    stream,
                    binary
                    ]),

        {ok, #state{
                pid = Pid,
                port = Port,
                s = Socket
                }};
    Error ->
        {stop, Error}
end.
~~~

To summarize:

* I think getting an 'EXIT' should force the gen_server to call terminate
  immediately by returning:

~~~ erlang
{stop, shutdown, State}
~~~

* the catch in terminate/2 should be used only for `erlang:port_close(Port)`

* we should fix the possible fd leak in init/2

What do you think?

Great find, thanks for figuring this all out!

**Comment by ksallberg** [comment]
I agree with the fd leak fixes.

> If we shutdown the gen_server using 'shutdown' or 'normal', it should
> suppress the error messages:
>     {stop, shutdown, State}

I tried this, however, returning that tuple (with shutdown, and normal) raises errors like:
```
{noproc,{gen_server,call,[<0.230.0>,stop]}}
{shutdown,{gen_server,call,[<0.244.0>,stop]}}
{noproc,{gen_server,call,[<0.454.0>,stop]}}
{normal,{gen_server,call,[<0.858.0>,stop]}}
```

Expanded error messages of the above:
```
12:57:15.903 [error] CRASH REPORT Process erlvirt with 2 neighbours exited with reason: {{noproc,{gen_server,call,[<0.131.0>,stop]}},{gen_server,call,[my,function,infinity]}} in gen_server:call/3 line 199
12:57:15.905 [error] Supervisor erlvirt_sup had child erlvirt started with erlvirt:start_link() at <0.86.0> exit with reason {{noproc,{gen_server,call,[<0.131.0>,stop]}},{gen_server,call,[my,function,infinity]}} in context child_terminated


12:57:15.902 [error] CRASH REPORT Process my with 0 neighbours exited with reason: no such process or port in call to gen_server:call(<0.131.0>, stop) in gen_server:terminate/6 line 760
12:57:15.902 [error] Supervisor erlvirt_sup had child domain started with my:start_link() at <0.85.0> exit with reason no such process or port in call to gen_server:call(<0.131.0>, stop) in context child_terminated
```
This leads me to believe that the process actually dies when reaching the handle_info for 'EXIT'.
I did not encounter any errors in /var/log/libvirt/libvirtd.log. However, I have observed that when the error (1) appears, then there is always something like *Closed fd 26* close in time. This is shown here:

```
2015-07-31 11:08:21.080+0000: 10448: debug : virConnectGetConfigFile:920 : Loading config file '/etc/libvi
rt/libvirt.conf'
2015-07-31 11:08:21.081+0000: 10448: debug : virConfReadFile:761 : filename=/etc/libvirt/libvirt.conf
2015-07-31 11:08:21.081+0000: 10448: debug : virFileClose:90 : Closed fd 26
2015-07-31 11:08:21.081+0000: 10448: debug : virConnectGetDefaultURI:1028 : Falling back to uri qemu:///system
2015-07-31 11:08:21.081+0000: 10448: debug : do_open:1103 : name "qemu:///system" to URI components:
  scheme qemu
  server <null>
  user <null>
  port 0
  path /system

2015-07-31 11:08:21.081+0000: 10448: debug : do_open:1149 : trying driver 0 (Test) ...
2015-07-31 11:08:21.081+0000: 10448: debug : do_open:1156 : driver 0 Test returned DECLINED
2015-07-31 11:08:21.081+0000: 10448: debug : do_open:1149 : trying driver 1 (OPENVZ) ...
2015-07-31 11:08:21.081+0000: 10448: debug : do_open:1156 : driver 1 OPENVZ returned DECLINED
2015-07-31 11:08:21.081+0000: 10448: debug : do_open:1149 : trying driver 2 (VMWARE) ...
2015-07-31 11:08:21.081+0000: 10448: debug : do_open:1156 : driver 2 VMWARE returned DECLINED
```

The above seems to have caused this specific error (log time +2 hours apparently):
```
13:08:21.088 [error] Supervisor erlvirt_sup had child domain started with my:start_link() at <0.137.0> exit with reason {shutdown,{gen_server,call,[<0.482.0>,stop]}} in context child_terminated
```

So, if the process dies, perhaps it is OK to return {noreply, State} to the caller of handle_info 'EXIT'?

**Comment by msantos** [comment]
1. handling 'EXIT'
~~~
12:57:15.902 [error] CRASH REPORT Process my with 0 neighbours exited with reason: no such process or port in call to gen_server:call(<0.131.0>, stop) in gen_server:terminate/6 line 760
12:57:15.902 [error] Supervisor erlvirt_sup had child domain started with my:start_link() at <0.85.0> exit with reason no such process or port in call to gen_server:call(<0.131.0>, stop) in context child_terminated
~~~
Ok, so it looks like what is happening is this:
* the port exits with 'EXIT'
* the gen_server stops and terminates
* erlvirt_sup calls verx_client:stop/1 but the gen_server has already exited

So I think verx_client:stop/1 should always return ok:
~~~ erlang
diff --git a/src/verx_client.erl b/src/verx_client.erl
index d5f76ff..91fc135 100644
--- a/src/verx_client.erl
+++ b/src/verx_client.erl
@@ -73,7 +73,8 @@ start_link(Arg) ->
     gen_server:start_link(Transport, [Self, Arg], []).

 stop(Ref) when is_pid(Ref) ->
-    gen_server:call(Ref, stop).
+    catch gen_server:call(Ref, stop),
+    ok.


 call(Ref, Proc) ->
~~~

2. Thanks for the debug log! So that confirms the theory that libvirtd is closing the socket.

Do you any limits set in libvirtd.conf like max_clients?

**Comment by ksallberg** [comment]
That makes sense. So we change verx_client:stop like you did there, and then keep {stop, shutdown, Status} in verx_client_unix? I could do that tomorrow

I had the default commented value of:
```
#max_clients = 20
```
changed it to 1000 but the error keeps appearing.

**Comment by msantos** [comment]
Sure, if you could make the changes on your branch, that would be great!

Do you have 10 persistent connections or are you cycling through connections?

You might need to change a few of the other settings (max_workers, max_requests, max_client_requests).

**Comment by ksallberg** [comment]
Yes I will do that. I am cycling through connections. Thanks for the suggestions, will look into them!

**Comment by ksallberg** [comment]
Ok I think I got the git squash command right this time ;) Let me know if something is missing.

**Comment by msantos** [comment]
Perfect, thanks Kristian!
